infrastructure/src/main/java/org/corfudb/infrastructure/management/ClusterGraph.java

```diff
@@ -0,0 +1,216 @@
+    }
+
+    /**
+     * Convert a cluster graph which could have asymmetric failures to a graph with symmetric failures between nodes.
+     *
+     * @return a graph with symmetric failures between nodes.
+     */
+    public ClusterGraph toSymmetric() {
+        Map<String, NodeConnectivity> symmetric = new HashMap<>();
+
+        graph.keySet().forEach(node -> {
+            NodeConnectivity adjacent = graph.get(node);
+            if (adjacent.getType() == NodeConnectivityState.UNAVAILABLE) {
+                symmetric.put(node, adjacent);
+                return;
+            }
+
+            ImmutableMap.Builder<String, Boolean> newConnectivity = ImmutableMap.builder();
+            adjacent.getConnectivity().keySet().forEach(adjNode -> {
+                boolean status = getConnectionStatus(node, adjNode);
+                boolean oppositeStatus = getConnectionStatus(adjNode, node);
+
+                newConnectivity.put(adjNode, status && oppositeStatus);
+            });
+
+            NodeConnectivity symmetricConnectivity = NodeConnectivity.builder()
+                    .endpoint(node)
+                    .connectivity(newConnectivity.build())
+                    .build();
+
+            symmetric.put(node, symmetricConnectivity);
+        });
+
+        return ClusterGraph.builder().graph(ImmutableMap.copyOf(symmetric)).build();
+    }
+
+    /**
+     * Get a decision maker node to detect a failure. It must have:
+     * - highest number of successful connections in the graph
+     * - quorum of connected nodes to be able to make a decision (to make a node failed)
+     *
+     * @return sorted set of nodes
+     */
+    public Optional<NodeRank> getDecisionMaker() {
+        log.trace("Get decision maker");
+
+        NavigableSet<NodeRank> nodes = getNodeRanks();
+
+        if (nodes.isEmpty()) {
+            log.error("Empty graph. Can't provide decision maker");
+            return Optional.empty();
+        }
+
+        int quorum = graph.size() / 2 + 1;
+        NodeRank first = nodes.first();
+
+        if (first.numConnections < quorum) {
+            log.error("No quorum to detect failed servers. Graph: {}, decision maker candidate: {}", this, first);
+            return Optional.empty();
+        }
+
+        log.trace("Decision maker has found: {}", first);
+        return Optional.of(first);
+    }
+
+    public Optional<NodeRank> findFailedNode() {
+        log.trace("Looking for failed node");
+
+        NavigableSet<NodeRank> nodes = getNodeRanks();
+        if (nodes.isEmpty()) {
+            log.error("Empty graph. Can't provide failed node");
+            return Optional.empty();
+        }
+
+        NodeRank last = nodes.last();
+        if (last.numConnections == graph.size()) {
+            return Optional.empty();
+        }
+
+        return Optional.of(last);
+    }
+
+    public ImmutableMap<String, NodeRank> findFullyConnectedResponsiveNodes(List<String> unresponsiveNodes) {
```

Comment:
Why return an ImmutableMap?

Reply:
- Prevent changing the map in other places.
- if a developer can't change the map then it pushes him to reorder his code according to this fact. In the other case, the developer just changes the map and it makes the algorithm more complicated. So it makes developers think about their workflow and write simple code not overcomplicate it, have/expose the workflow in compile time not in runtime